tests/sles4sap/publiccloud/qesap_terraform.pm

```diff
@@ -95,34 +90,76 @@ sub create_hana_vars_section {
     my %hana_vars;
     if ($ha_enabled == 1) {
-        $hana_vars{sap_hana_install_software_directory} = get_required_var("HANA_MEDIA");
-        $hana_vars{sap_hana_install_master_password} = get_required_var("_HANA_MASTER_PW");
-        $hana_vars{sap_hana_install_sid} = get_required_var("INSTANCE_SID");
-        $hana_vars{sap_hana_install_instance_number} = get_required_var("INSTANCE_ID");
-        $hana_vars{sap_domain} = get_var("SAP_DOMAIN", "qesap.example.com");
-        $hana_vars{primary_site} = get_var("HANA_PRIMARY_SITE", "site_a");
-        $hana_vars{secondary_site} = get_var("HANA_SECONDARY_SITE", "site_b");
-        set_var("SAP_SIDADM", lc(get_var("INSTANCE_SID") . "adm"));
+        $hana_vars{sap_hana_install_software_directory} = get_required_var('HANA_MEDIA');
+        $hana_vars{sap_hana_install_master_password} = get_required_var('_HANA_MASTER_PW');
+        $hana_vars{sap_hana_install_sid} = get_required_var('INSTANCE_SID');
+        $hana_vars{sap_hana_install_instance_number} = get_required_var('INSTANCE_ID');
+        $hana_vars{sap_domain} = get_var('SAP_DOMAIN', 'qesap.example.com');
+        $hana_vars{primary_site} = get_var('HANA_PRIMARY_SITE', 'site_a');
+        $hana_vars{secondary_site} = get_var('HANA_SECONDARY_SITE', 'site_b');
+        set_var('SAP_SIDADM', lc(get_var('INSTANCE_SID') . 'adm'));
     }
     return (\%hana_vars);
 }
 
+=head2 create_instance_data
+
+    Create and populate a publiccloud::provider compatible memory structure with the current
+    deployment instances.
+
+=cut
+
+sub create_instance_data {
+    my $provider = shift;
+    my $class = ref($provider);
+    die "Unexpected class type [$class]" unless ($class =~ /^publiccloud::(azure|ec2|gce)/);
+    my @instances = ();
+    my $inventory_file = qesap_get_inventory(get_required_var('PUBLIC_CLOUD_PROVIDER'));
+    my $ypp = YAML::PP->new;
+    my $raw_file = script_output("cat $inventory_file");
+    my $inventory_data = $ypp->load_string($raw_file)->{all}{children};
+
+    for my $type_label (keys %$inventory_data) {
+        my $type_data = $inventory_data->{$type_label}{hosts};
+        for my $vm_label (keys %$type_data) {
+            my $instance = publiccloud::instance->new(
+                public_ip => $type_data->{$vm_label}->{ansible_host},
+                instance_id => $vm_label,
+                username => get_required_var('PUBLIC_CLOUD_USER'),
+                ssh_key => '~/.ssh/id_rsa',
+                provider => $provider,
+                region => $provider->provider_client->region,
+                type => get_required_var('PUBLIC_CLOUD_INSTANCE_TYPE'),
+                image_id => $provider->get_image_id()
+            );
+            push @instances, $instance;
+        }
+    }
+    publiccloud::instances::set_instances(@instances);
+    return \@instances;
+}
+
 sub run {
     my ($self, $run_args) = @_;
+    my $workspace = get_var('PUBLIC_CLOUD_RESOURCE_GROUP', 'qesapopenqa');
```

Comment:
Variable renaming proposal: `$workspace` --> `$resource_group`

Reply:
It's a resource group and a workspace _in Azure_, but in GCE and EC2 it's only used as a workspace.

I read this line as: "let's define a workspace; if we have a value for PUBLIC_CLOUD_RESOURCE_GROUP, we will use that as a workspace name, otherwise we will use 'qesapopenqa'", so IMO it's a better fit to call the variable `$workspace`.

I can add a comment before L144 to make this clearer.